c7n/resources/s3.py

```diff
@@ -3423,3 +3423,99 @@ def process_bucket(self, bucket):
+VALUE_FILTER_MAGIC_VALUES = ['absent', 'present', 'not-null', 'empty']
+
+
+@filters.register('ownership')
+class BucketOwnershipControls(BucketFilterBase, ValueFilter):
+    """Filter for object ownership controls
+
+    Reference: https://docs.aws.amazon.com/AmazonS3/latest/userguide/about-object-ownership.html
+
+    :example
+
+    Find buckets with ACLs disabled
+
+    .. code-block:: yaml
+
+            policies:
+              - name: s3-bucket-acls-disabled
+                resource: aws.s3
+                region: us-east-1
+                filters:
+                  - type: ownership
+                    value: BucketOwnerEnforced
+
+    :example
+
+    Find buckets with object ownership preferred or enforced
+
+    .. code-block:: yaml
+
+            policies:
+              - name: s3-bucket-ownership-preferred
+                resource: aws.s3
+                region: us-east-1
+                filters:
+                  - type: ownership
+                    op: in
+                    value:
+                      - BucketOwnerEnforced
+                      - BucketOwnerPreferred
+
+    :example
+
+    Find buckets with no object ownership controls
+
+    .. code-block:: yaml
+
+            policies:
+              - name: s3-bucket-no-ownership-controls
+                resource: aws.s3
+                region: us-east-1
+                filters:
+                  - type: ownership
+                    value: empty
+    """
+    schema = type_schema('ownership', rinherit=ValueFilter.schema, value={'oneOf': [
+        {'type': 'string', 'enum': OWNERSHIP_CONTROLS + VALUE_FILTER_MAGIC_VALUES},
+        {'type': 'array', 'items': {
+            'type': 'string', 'enum': OWNERSHIP_CONTROLS + VALUE_FILTER_MAGIC_VALUES}}]})
+    permissions = ('s3:GetBucketOwnershipControls',)
+    annotation_key = 'c7n:ownership'
+
+    def __init__(self, data, manager=None):
+        super(BucketOwnershipControls, self).__init__(data, manager)
+
+        # Ownership controls appear as an array of rules. There can only be one
+        # ObjectOwnership rule defined for a bucket, so we can automatically
+        # match against that if it exists.
+        self.data['key'] = f'("{self.annotation_key}".Rules[].ObjectOwnership)[0]'
+
+    def process(self, buckets, event=None):
+        with self.executor_factory(max_workers=2) as w:
+            futures = {w.submit(self.process_bucket, b): b for b in buckets}
+            for future in as_completed(futures):
+                b = futures[future]
+                if future.exception():
+                    self.log.error("Message: %s Bucket: %s", future.exception(),
+                                   b['Name'])
+                    continue
+        return super(BucketOwnershipControls, self).process(buckets, event)
+
+    def process_bucket(self, b):
+        client = bucket_client(local_session(self.manager.session_factory), b)
+        if self.annotation_key not in b:
```

Comment:
nit: most of the block can be dendented if the condition is flipped `if self.annotation_key in b: return`

Reply:
Oh yeah that's much cleaner. I was also setting a variable for no reason in the `else`, copypasta fail :grin: . Thanks :+1: